aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteModelTree.java

```diff
@@ -0,0 +1,157 @@
+import com.amplifyframework.util.Empty;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Utility class to help traverse a tree of models by relationship.
+ */
+final class SQLiteModelTree {
+    private final ModelSchemaRegistry registry;
+    private final SQLCommandFactory commandFactory;
+    private final SQLiteDatabase database;
+
+    /**
+     * Constructs a model family tree traversing utility.
+     * @param registry model registry to search schema from
+     * @param commandFactory SQL command factory
+     * @param database SQLite database connection handle
+     */
+    SQLiteModelTree(ModelSchemaRegistry registry,
+                    SQLCommandFactory commandFactory,
+                    SQLiteDatabase database) {
+        this.registry = registry;
+        this.commandFactory = commandFactory;
+        this.database = database;
+    }
+
+    /**
+     * Returns a map of descendants of a set of models (of same type).
+     * A model is a child of its parent if it uses its parent's ID as foreign key.
+     * @param root Collection of models to query its descendants of.
+     * @return Map of descendants keyed by model schema. The value contains a set of
+     *          descendants' IDs for that model type.
+     */
+    <T extends Model> Map<ModelSchema, Set<String>> descendantsOf(Collection<T> root) {
+        if (Empty.check(root)) {
+            throw new IllegalArgumentException("Cannot traverse tree from an empty root.");
+        }
+        Map<ModelSchema, Set<String>> descendants = new LinkedHashMap<>();
+        ModelSchema rootSchema = registry.getModelSchemaForModelInstance(root.iterator().next());
+        Set<String> rootIds = new HashSet<>();
+        for (T model : root) {
+            rootIds.add(model.getId());
+        }
+        recurseTree(descendants, rootSchema, rootIds);
+        return descendants;
+    }
+
+    private void recurseTree(
+            Map<ModelSchema, Set<String>> tree,
+            ModelSchema modelSchema,
+            Collection<String> parentIds
+    ) {
+        SQLiteTable parentTable = SQLiteTable.fromSchema(modelSchema);
+        for (ModelAssociation association : modelSchema.getAssociations().values()) {
+            switch (association.getName()) {
+                case "HasOne":
+                case "HasMany":
+                    String childModel = association.getAssociatedType(); // model name
+                    ModelSchema childSchema = registry.getModelSchemaForModelClass(childModel);
+                    SQLiteTable childTable = SQLiteTable.fromSchema(childSchema);
+                    String childPrimaryKey = childTable.getPrimaryKey().getAliasedName();
+                    QueryField queryField = QueryField.field(parentTable.getPrimaryKeyColumnName());
+
+                    // Chain predicates with OR operator.
```

Comment:
Can you add a test that exercises this predicate building logic?  I think your current test is too simple to verify it.   Maybe add another test that deletes a `Blog`, which has 3 `Post`s, each of which has 3 `Comment`s, and then verify that a total of 13 `StorageItemChange` items are observed?